### ISSUE TITLE
Inconvenient import suggestion order 

### ISSUE THREAD

**Issue by incendial** [Dart-Code/Dart-Code#4522]
**Describe the bug**
I have a function with a similar name in several modules + have a lint rule for relative imports.
The problems is that the expected import suggestion somehow appears at the bottom of the suggestion list (I was expecting it to be the first one). Here is an example:
<img width="553" alt="Screenshot 2023-05-05 at 14 24 42" src="https://user-images.githubusercontent.com/21116742/236434330-066fe836-ee9b-4f66-9d44-a1f6c5ff9c17.png">

The one I need is actually **'reporters/reporter_factory.dart'**.

**To Reproduce**
Here is an example project stricture https://github.com/dart-code-checker/dart-code-metrics/tree/master/lib/src/analyzers. 
As you can see, each subfolder has a folder named reporters with a 'reporter_factory.dart' file, that declares a 'reporter' function.
To repro the problem, try to reimport the 'reporter' in any of the analyzers.

**Please complete the following information:**
 - Operating System and version: macOS
 - VS Code version: Version: 1.78.0
 - Dart extension version: v3.63.20230502
 - Dart/Flutter SDK version: dart 2.19.6

**Comment by incendial** [comment]
> I can think of some rules that might fix this (eg. preferring imports that don't start with ../, or the ones with the last path segments) but I don't know if they're clearly better to make the standard behaviour?

From my perspective, if the relative imports rule is enabled, the sorting should consider placing the shortest entries (by path segments and then by characters) first. I don't see any strait-forward case when this rule won't work. 

> The idea of using ../ as a feature is interesting, but I'm wondering whether that's really the feature that matters or whether it might not be better to focus on the 'distance' to the file (counting each / as a distance of 1). It seems to me that the closer the files are on disk the more likely it is to be the right definition. It would certainly have produced the desired outcome in this case; what I don't know is whether there are counter examples.

So, yeah, Brian described this way better than me.

**Comment by incendial** [comment]
> As of this writing we haven't really put any thought into prioritizing fixes because we've had few enough that it hasn't become an issue until now.

Any chance after dealing with this you might want to reconsider the "Add all missing imports" fix, considering, that the list will be ordered and the first import would be more likely the desired one?

**Comment by DanTup** [comment]
> The idea of using ../ as a feature is interesting, but I'm wondering whether that's really the feature that matters or whether it might not be better to focus on the 'distance' to the file

Yep, I think so. That's what "or the ones with the last path segments" was intended to mean, but I typo'd "least" 😄

I'll open a CL to fix the sorting in LSP (eg., sort by priority then index, not priority then title). If we think alphabetical makes sense in some cases, then the fix producer can always sort them that way, but having LSP re-sort them when it has much less context doesn't seem to make sense. It appears I added it as part of https://github.com/dart-lang/sdk/commit/00cc25a28b6c4af5613887256448156e86a654ee but it's not clear this was necessary for the fix it was making 🤷‍♂️

> if the user has enabled a lint to prefer one kind of import over another, then I think that we should only suggest the fixes that match that lint

That sounds reasonable to me. Currently we only track one lint (`CodeStyleOptions.useRelativeUris` from `prefer_relative_imports`) but we do have an opposite lint (`always_use_package_imports`) that we could use to make this tri-state (no preference, prefer relative, prefer absolute).


For the sorting, I wonder whether having two levels of ranking will work (eg. the existing ranking on FixKind, and then something else that's specific to each fix), or whether there might be a case where different fix-kinds could be interleaved? Does two fixes having the same kind indicate that they have approximately the same relevance?


> Any chance after dealing with this you might want to reconsider the "Add all missing imports" fix, considering, that the list will be ordered and the first import would be more likely the desired one?

I'm not sure anyone is against doing this - but I think having the options sorted well is a prerequisite for it to be useful.

**Comment by incendial** [comment]
> If we think alphabetical makes sense in some cases, then the fix producer can always sort them that way

But how about sorting by the length as well? For example "relative/to/my/file.dart" and "relative/to/my/another_file.dart" have the same segments number, but I'd be happy to see predictable outcome here. It also allows to pick faster due to predictability.

> but I think having the options sorted well is a prerequisite for it to be useful.

👍

**Comment by DanTup** [comment]
This also came up at https://github.com/Dart-Code/Dart-Code/issues/5838

The request is for it to be customisable, but I wonder if just improving the standard behaviour is better (as far as I can tell, the priority is identical regardless of whether it's an external package or the local package).

<img width="1728" height="848" alt="Image" src="https://github.com/user-attachments/assets/32bfb13c-19de-40cd-9b7c-903e65ae404c" />

**Comment by bwilkerson** [comment]
We aren't likely to add customization options until we've first investigated improving the order without them and determined that there is no generally 'right' ordering. And I'm not sure we'd do so even then.
